### PR TITLE
LoRA+ support

### PR DESCRIPTION
**Pull request by kallewoof**
Some issues that need addressing, so keeping as draft.

- [ ] Deal with model params being passed around and then discarded when LoRA+ is used. Why are these even needed in DeepSpeed space? Can we just `None` them, or at least `None` them when LoRA+ is activated?
- [x] Wait for merge: https://github.com/huggingface/peft/pull/1973
- [x] Wait for next PEFT release

### CODE REVIEW

**Comment by tdrussell** [comment]
This looks good. Have you tested it yet? One thing to check is that the lora+ optimizer parameters get set correctly. The create_loraplus_optimizer() takes the model instead of the parameters, presumably internally it is still only grabbing the parameters that requires_grad? Need to make sure of that.

**Comment by kallewoof** [comment]
> The create_loraplus_optimizer() takes the model instead of the parameters, presumably internally it is still only grabbing the parameters that requires_grad? Need to make sure of that.

It does that, yes

https://github.com/huggingface/peft/blob/b9260305e3e0129ba35214a40ca7ce60d7ea25b3/src/peft/optimizers/loraplus.py#L70-L83

and it also does bnb module override registration stuff if it is able to

https://github.com/huggingface/peft/blob/b9260305e3e0129ba35214a40ca7ce60d7ea25b3/src/peft/optimizers/loraplus.py#L70-L83

**Comment by kallewoof** [comment]
> Have you tested it yet? 

I have used this exclusively since I made the pull request, a couple of fine tunes so far, and working on a 70B fine tune right now. The results look fine to me so far, but I will know more after spending some time with it.

**Comment by kallewoof** [comment]
Training broke down when I tried using this at the recommended ratio 16. Testing other (lower) ratio values to see if they work better.
At ratio 5, the training seemed to work, but with lora+ disabled, the training seemed to give slightly better eval loss during the initial stages of training. Will experiment with this some more and do full training runs comparing eval with lora+ disabled and with different ratios.

**Comment by kallewoof** [comment]
Closing this. Will look into it again when the next PEFT has been released, if I find the time.

**Comment by kallewoof** [comment]
I have been training with this on and off for awhile and it seems to work as it should, so reopening this.

**Comment by jukofyork** [comment]
Just having a look at this now and wondered where is gets the `lr` value from when the parameter isn't passed:

https://github.com/huggingface/peft/blob/main/src/peft/optimizers/loraplus.py

```python
    kwargs["lr"] = lr
    loraplus_weight_decay = kwargs.pop("loraplus_weight_decay", 0.0)
    loraplus_lr_embedding = kwargs.pop("loraplus_lr_embedding", 1e-6)

    optimizer_grouped_parameters = [
        {
            "params": list(param_groups["groupA"].values()),
            "weight_decay": loraplus_weight_decay,
            "lr": lr,
        },
        {
            "params": list(param_groups["embedding"].values()),
            "weight_decay": loraplus_weight_decay,
            "lr": loraplus_lr_embedding,
        },
        {
            "params": list(param_groups["groupB"].values()),
            "weight_decay": loraplus_weight_decay,
            "lr": lr * loraplus_lr_ratio,
        },
        {
            "params": list(param_groups["groupB_no_decay"].values()),
            "weight_decay": 0.0,
            "lr": lr * loraplus_lr_ratio,
        },
    ]
```

and also if it overwriting the `weight_decay` value with zero?

**Comment by jukofyork** [comment]
I can's see how `kwargs["lr"] = lr` works if you don't pass it? Does it get it from the `optimizer_kwargs` parameters?

(sorry for my python ignorance - this just seems really alien to me!)

**Comment by jukofyork** [comment]
> Training broke down when I tried using this at the recommended ratio 16. Testing other (lower) ratio values to see if they work better. At ratio 5, the training seemed to work, but with lora+ disabled, the training seemed to give slightly better eval loss during the initial stages of training. Will experiment with this some more and do full training runs comparing eval with lora+ disabled and with different ratios.

@kallewoof 

How quickly did the training break down when using 16x?

Also could the differences be due to losing the `weight_decay` setting and having it reset to zero?

**Comment by kallewoof** [comment]
> Just having a look at this now and wondered where it gets the `lr` value from when the parameter isn't passed:
> 
> ```python
>         if optim_config.get('use_loraplus', False):
>             loraplus_lr_ratio = optim_config.get('loraplus_lr_ratio', 16)
>             # TODO: handle params being thrown out here; why is it included in the first place?
>             # delete 'params' from optimizer_kwargs
>             del optimizer_kwargs['params']
>             return create_loraplus_optimizer(
>                 model=pipeline_model,
>                 optimizer_cls=optimizer_cls,
>                 loraplus_lr_ratio=loraplus_lr_ratio,
>                 **optimizer_kwargs
>             )
> ```

> I can's see how `kwargs["lr"] = lr` works if you don't pass it? Does it get it from the `optimizer_kwargs` parameters?
> 
> (sorry for my python ignorance - this just seems really alien to me!)

It is passed in `**optimizer_kwargs` which has it set in

```python
    def get_optimizer(model_parameters):
        optim_type = optim_config['type'].lower()
        optimizer_kwargs = {
            "params": model_parameters,
            "lr": lr,        #        <<============== here
            "betas": (optim_config.get('beta1', 0.9), optim_config.get('beta2', 0.99)),
            "weight_decay": optim_config.get('weight_decay', 0.01),
            "eps": optim_config.get('eps', 1e-6)
        }
```

> and also if it's overwriting the `weight_decay` value with zero?

It might, actually. Hm.

**Comment by jukofyork** [comment]
> > Just having a look at this now and wondered where it gets the `lr` value from when the parameter isn't passed:
> > ```python
> >         if optim_config.get('use_loraplus', False):
> >             loraplus_lr_ratio = optim_config.get('loraplus_lr_ratio', 16)
> >             # TODO: handle params being thrown out here; why is it included in the first place?
> >             # delete 'params' from optimizer_kwargs
> >             del optimizer_kwargs['params']
> >             return create_loraplus_optimizer(
> >                 model=pipeline_model,
> >                 optimizer_cls=optimizer_cls,
> >                 loraplus_lr_ratio=loraplus_lr_ratio,
> >                 **optimizer_kwargs
> >             )
> > ```
> 
> > I can's see how `kwargs["lr"] = lr` works if you don't pass it? Does it get it from the `optimizer_kwargs` parameters?
> > (sorry for my python ignorance - this just seems really alien to me!)
> 
> It is passed in `**optimizer_kwargs` which has it set in
> 
> ```python
>     def get_optimizer(model_parameters):
>         optim_type = optim_config['type'].lower()
>         optimizer_kwargs = {
>             "params": model_parameters,
>             "lr": lr,        #        <<============== here
>             "betas": (optim_config.get('beta1', 0.9), optim_config.get('beta2', 0.99)),
>             "weight_decay": optim_config.get('weight_decay', 0.01),
>             "eps": optim_config.get('eps', 1e-6)
>         }
> ```

Ah, thanks - pythonisms like this seem to make zero sense to me! :)
